numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,168 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include <Python.h>
+
+#include <unordered_map>
+#include <functional>
+
+#include "numpy/arrayobject.h"
+#include "numpy/npy_common.h"
+
+
+template<typename T>
+PyObject* unique(PyArrayObject *self)
+{
+    /* This function takes a numpy array and returns a numpy array containing
+    the unique values.
+
+    It assumes the numpy array includes data that can be viewed as unsigned integers
+    of a certain size (sizeof(T)).
+
+    It doesn't need to know the actual type, since it needs to find unique values
+    among binary representations of the input data. This means it won't apply to
+    custom or complicated dtypes or string values.
+    */
+    NpyIter* iter;
+    NpyIter_IterNextFunc *iternext;
+    char** dataptr;
+    npy_intp* strideptr,* innersizeptr;
+    std::unordered_map<T, char> hashmap;
+
+    iter = NpyIter_New(self, NPY_ITER_READONLY|
+                             NPY_ITER_EXTERNAL_LOOP|
+                             NPY_ITER_REFS_OK,
+                        NPY_KEEPORDER, NPY_NO_CASTING,
+                        NULL);
+    if (iter == NULL) {
+        return Py_None;
+    }
+
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        NpyIter_Deallocate(iter);
+        return Py_None;
+    }
+    dataptr = NpyIter_GetDataPtrArray(iter);
+    strideptr = NpyIter_GetInnerStrideArray(iter);
+    innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    // first we put the data in a hash map
+    do {
+        char* data = *dataptr;
+        npy_intp stride = *strideptr;
+        npy_intp count = *innersizeptr;
+
+        while (count--) {
+            hashmap[*((T *) data)] = 0;
+            data += stride;
+        }
+    } while(iternext(iter));
+    NpyIter_Deallocate(iter);
+
+    // then we iterate through the map's keys to get the unique values
+    T* res = new T[hashmap.size()];
+    auto it = hashmap.begin();
+    size_t i = 0;
+    for (; it != hashmap.end(); it++, i++) {
+        res[i] = it->first;
+    }
+
+    // does this need to have the same lifetime as the array?
+    npy_intp dims[1] = {(npy_intp)hashmap.size()};
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    Py_INCREF(descr);
+    return PyArray_NewFromDescr(
+        &PyArray_Type,
+        descr,
+        1, // ndim
+        dims, // shape
+        NULL, // strides
+        res, // data
+        NPY_ARRAY_OUT_ARRAY, // flags
+        NULL // obj
+    );
+}
+
+
+// this map contains the functions used for each item size.
+typedef std::function<PyObject *(PyArrayObject *)> function_type;
+std::unordered_map<size_t, function_type> unique_funcs = {
+    {sizeof(npy_uint8), unique<npy_uint8>},
+    {sizeof(npy_uint16), unique<npy_uint16>},
+    {sizeof(npy_uint32), unique<npy_uint32>},
+    {sizeof(npy_uint64), unique<npy_uint64>},
+};
```

Comment:
If you plug this into the main `unique` function, you should find some errors.  The point being: Maybe best not to try to be smart (bools are also a slight problem).

For floats the hash function has to account for `0` and `-0` being different (the NaN behavior may be trickier, since I am not sure C++ hashmap accounts for it, even if the float hashing does, does the comparison use something NaN aware?).

This all seems light-weight enough that I am happy to go there (not sure if unordered map is ideal, but probably good enough in practice; for one I am not sure if unordered map actually does hash randomization).

I am always a bit concerned that I would like to be able to provide a public API for others to implement this for their DTypes, and one that ideally doesn't use any NumPy array at all.  I think that would mean splitting this into 3 functions:
* One function which sets up the iterator and does the outer iteration (the `iternext`).  It then would call, unfortunately maybe 3 other functions (although that can all just be templated for now I think):
  * Create a new hash-map of the right type. 
  * Call one function which just deals with updating the hash-map for the 1-D slice (i.e. that `while (count--)` loop).
  * Split out the finalization of the hash-map into a NumPy array (or a single contiguous allocation).

If we had those, it should be straight forward to make them available to user DTypes.  Of course such a reorganization isn't terrible hard even if we don't do it now.

---

This code also needs handling with C++ exceptions (at the very least memory errors), so that is a bit of a problem, we may need to steal a generic translator (e.g. from `pybind11`), so that you can do `translate_errors_cpp_meth<PyArray_Unique>` when registering below.

Reply:
Actually, strike that: There is a bigger problem with C++ exceptions: You would have to use a helper `OwnedRef` object in all such code to ensure you clean up correctly.
So catching is only straight forward if no `PyObject` or manual memory handling is happening inside the functions being caught.